### PR TITLE
Protect against a non-privileged port connecting to us when we are running as root

### PR DESCRIPTION
**Pull request by rhc54**
@jsquyres Mind just verifying the logic on this one? Plugs a potential security hole causing angst at the mother ship

### CODE REVIEW

**Comment by jsquyres** [comment]
@rhc54 It looks like you're checking the file descriptor number, not the port number.

**Comment by rhc54** [comment]
@jsquyres Could you take a quick gander? Turns out you cannot use getpeername as the sockets have to be connected first, and that is what we are trying to avoid. However, I am able to get the port number of the inbound connection once we accept it, and then can break things at that time.

File: orte/mca/oob/tcp/oob_tcp_listener.c [hunk 1]
```diff
@@ -909,7 +936,9 @@ static void connection_event_handler(int incoming_sd, short flags, void* cbdata)
 static void tcp_ev_cons(mca_oob_tcp_listener_t* event)
 {
     event->ev_active = false;
+    event->tcp6 = false;
     event->sd = -1;
+    event->port = -1;
```
Comment (jsquyres) [inline orte/mca/oob/tcp/oob_tcp_listener.c]: The port is a `uint16_t` -- this assignment may well generate a compiler warning...?

**Comment by jsquyres** [comment]
Otherwise :+1:

**Comment by rhc54** [comment]
Good point about the default value for the port number - I changed it to zero. Will let this run thru the Jenkins tests and commit. Thanks!

**Comment by rhc54** [comment]
@bosilca would you mind taking a quick glance at this? The goal is to try and protect against connections from non-privileged ports if we are operating at a privileged level with a privileged port. I think I'm checking the remote port properly, but would welcome another set of eyes to ensure it's doing what I expected.

File: orte/mca/oob/tcp/oob_tcp_listener.c [hunk 2]
```diff
@@ -739,6 +742,30 @@ static void* listen_thread(opal_object_t *obj)
                 pending_connection->fd = accept(sd,
                                                 (struct sockaddr*)&(pending_connection->addr),
                                                 &addrlen);
+
+                /* if we are on a privileged port, we only accept connections
+                 * from other privileged sockets. A privileged port is one
+                 * whose fd is less than 1024 on Linux, so we'll check for that. */
```
Comment (ggouaillardet) [inline orte/mca/oob/tcp/oob_tcp_listener.c]: Two typos
"A privileged socket is one whose port is less than ..."

File: orte/mca/oob/tcp/oob_tcp_listener.c [hunk 3]
```diff
@@ -739,6 +742,30 @@ static void* listen_thread(opal_object_t *obj)
...
+                    }
+                    if (1024 < inport) {
+                        /* someone tried to cross-connect privileges,
+                         * say something */
+                        orte_show_help("help-oob-tcp.txt",
+                                       "privilege failure",
+                                       true, opal_process_info.nodename,
+                                       listener->port, inport);
+                        OBJ_RELEASE(pending_connection);
+                        CLOSE_THE_SOCKET(sd);
```
Comment (ggouaillardet) [inline orte/mca/oob/tcp/oob_tcp_listener.c]: Do you really want to close sd (aka the listening socket) ?
Or did you mean to close pending_connection->fd (aka the non privileged socket) ?

**Comment by rhc54** [comment]
@ggouaillardet Good points - thx!

File: orte/mca/oob/tcp/oob_tcp_listener.c [hunk 4]
```diff
@@ -739,8 +742,31 @@ static void* listen_thread(opal_object_t *obj)
...
+                        inport = ntohs(((struct sockaddr_in*)&pending_connection->addr)->sin_port);
+                    }
+                    if (1024 < inport) {
+                        /* someone tried to cross-connect privileges,
+                         * say something */
+                        orte_show_help("help-oob-tcp.txt",
+                                       "privilege failure",
+                                       true, opal_process_info.nodename,
+                                       listener->port, inport);
+                        OBJ_RELEASE(pending_connection);  // closes the incoming fd
```
Comment (ggouaillardet) [inline orte/mca/oob/tcp/oob_tcp_listener.c]: Do you intentionally __not__ close ```pending_connection->fd``` ?
Or should the destructor (to be created) do that for you ?

Comment (rhc54) [inline orte/mca/oob/tcp/oob_tcp_listener.c]: Sorry - was looking at the wrong destructor. No, it needs to close it.

**Pull request merged**